Tests to see if ideal in quaternion algebra is primitive (cyclic)

giacomoborin opened the pull request:
New methods for `QuaternionFractionalIdeal_rational`:

- `QuaternionFractionalIdeal_rational.is_integral()`
- `QuaternionFractionalIdeal_rational.is_primitive()`
- `QuaternionFractionalIdeal_rational.primitive_decomposition()`

These methods verify if an ideal in quaternion rational algebra is integral, primitive (cyclic) and decompose it as a primitive ideal. Part of the code came from [https://learningtosqi.github.io](https://learningtosqi.github.io). 
Also added the _Quaternion Algebras_ book by J. Voight in the bibliography. 

Done with @gioella

#sd123

giacomoborin committed:
add Quaternion Book to Reference

giacomoborin committed:
Add primitive decomposition and integrality check

giacomoborin committed:
Error in comments

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -3020,6 +3020,119 @@ def cyclic_right_subideals(self, p, alpha=None):
+        O_basis = self.left_order().basis_matrix()
+
+        # Write I in the basis of its left order via rref
+        M = block_matrix(1,2,[O_basis.transpose(),I_basis.transpose()]).rref()[:,4:]
```
I believe this is just `M = O_basis.solve_left(I_basis)`?

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -3020,6 +3020,119 @@ def cyclic_right_subideals(self, p, alpha=None):
+
+        # Write I in the basis of its left order via rref
+        M = block_matrix(1,2,[O_basis.transpose(),I_basis.transpose()]).rref()[:,4:]
+        g = Integer(gcd((gcd(M_row) for M_row in M)))
```
Shorter way: `g = Integer(gcd(M.list()))`.

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -3020,6 +3020,119 @@ def cyclic_right_subideals(self, p, alpha=None):
+        g = Integer(gcd((gcd(M_row) for M_row in M)))
+
+        # If g is 1 then the ideal is primitive
+        if g == 1:
```
The test `g.is_one()` is significantly faster than `g == 1`, since the latter goes through the type coercion system to find a common parent type to do the comparison in.

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -3020,6 +3020,119 @@ def cyclic_right_subideals(self, p, alpha=None):
+            return all([b in self.right_order() for b in self.basis()])
+        else:
+            self_square = self**2
+            return all([b in self for b in self_square.basis()])
```
Each of these tests should be equivalent to one of the form `A.free_module() <= B.free_module()`. (See also #37113.)

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -3020,6 +3020,119 @@ def cyclic_right_subideals(self, p, alpha=None):
+
+        """
+        _,g = self.primitive_decomposition()
+        return 1 == g
```
Same here, using `.is_one()` is faster.

giacomoborin committed:
format correction

giacomoborin committed:
Merge branch 'develop' into develop

giacomoborin committed:
Inserting reviews

In this small commit I inserted all the improvements pointed out previously by @yyyyx4 (many thanks!)

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -3059,6 +3059,114 @@ def cyclic_right_subideals(self, p, alpha=None):
+
+        TESTS::
+
+        Checks on random crafted ideals that they decompose as expected.
```
One more linter failure:

```suggestion
        TESTS:

        Checks on random crafted ideals that they decompose as expected::
```

(The `::` syntax only indicates that an indented block is about to follow, so here it should be moved down as indicated instead of sticking to the `TESTS:` header.)

yyyyx4 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -56,7 +56,7 @@
 from sage.structure.parent import Parent
 from sage.matrix.matrix_space import MatrixSpace
-from sage.matrix.constructor import diagonal_matrix, matrix
+from sage.matrix.constructor import diagonal_matrix, matrix, block_matrix
```
The import of `block_matrix` is now unused and can be removed.

giacomoborin committed:
removed block_matrix import

fchapoton committed:
fix doc syntax in quaternion_algebra.py
